ocfweb/docs/docs/staff/procedures/deploying-services.md

```diff
@@ -0,0 +1,379 @@
+[[!meta title="Deploying Services"]]
+
+## Overview
+Are you looking to deploy a new service to the OCF Kubernetes cluster? This
+document will cover all of the steps required to do so. Do note that this is **not** a
+substitute for a Kubernetes tutorial or a Docker tutorial _(there are many
+resources online for that)_ but a guide for getting your service running on the
+OCF. There are a lot of pieces of OCF infrastructure at play here, but this
+should serve as a gentle introduction into how things are done.
+
+This `HOWTO` uses one of the OCF's simplest services as an example:
+[Templates][templates]. Templates is service used internally by OCF staff
+serving 'copy-pasteable' email templates. You can use [[git|doc
+staff/backend/git]] to `clone` this [repo](https://github.com/ocf/templates/blob/master/kubernetes/templates.yml.erb)
+, or just follow along looking at the files. If you want to blaze ahead,
+this repo should serve as a good example of the required components in the
+repo for deployment.
+
+## Prerequisites
+If you're looking to deploy your service, you should first have a service to
+deploy. This is normally done with a repository underneath the
+[OCF's Github organization](https://github.com/ocf). If you have the correct
+permissions, you should be able to transfer, fork, or just create a
+repository in the org.
+
+### Docker & Jenkins
+
+For your service to be deployed onto the Kubernetes cluster, it must have
+a `Dockerfile` so that we can containerize the service. Nothing OCF specific
+here, just include whatever your service needs.
+
+[[Jenkins|doc staff/backend/jenkins]] will try to build that container,
+and then send the image to the OCF Docker server. This has to be specified in
+the `Makefile` at the root of the repository. Here's part of what templates has:
+
+```
+BIN := venv/bin
+
+DOCKER_REVISION ?= testing-$(USER)
+DOCKER_TAG = docker-push.ocf.berkeley.edu/templates:$(DOCKER_REVISION)
+RANDOM_PORT := $(shell expr $$(( 8000 + (`id -u` % 1000) + 1 )))
+
+
+.PHONY: dev
+dev: cook-image
+    @echo "Will be accessible at http://$(shell hostname -f ):$(RANDOM_PORT)/"
+    docker run --rm -p "$(RANDOM_PORT):8000" "$(DOCKER_TAG)"
+
+.PHONY: cook-image
+cook-image:
+    docker build --pull -t $(DOCKER_TAG) .
+
+.PHONY: push-image
+push-image:
+    docker push $(DOCKER_TAG)
+```
+
+The important targets to pay attention here are `cook-image` and `push-image`,
+Jenkins will try to run these as part of the deployment pipeline that continuously
+deploys to Kubernetes.
```

Comment:
`make test` is also a good one to mention, it's automatically run if it exists as part of the service pipeline so that's a great place to put any tests that a service may have.